parser2/src/grammar.lalrpop

```diff
@@ -0,0 +1,233 @@
+
+        "OpenBlock" => Token::OpenBlock,
+        "CloseBlock" => Token::CloseBlock,
+        "Separator" => Token::Separator,
+    }
+}
+
+// Utils
+
+Comma<Rule>: Vec<Rule> =
+    <rules: (<Rule> ",")*> <last: Rule?> => {
+        let mut rules = rules;
+        rules.extend(last);
+        rules
+    };
+
+And<Rule>: Vec<Rule> =
+    <rules: (<Rule> "and")*> <last: Rule> =>
+        rules.into_iter().chain(iter::once(last)).collect();
+
+pub Ident: Ident = {
+    "Name" => Ident::Name(<>.to_string()),
+    "(" <op: "Operator"> ")" => Ident::Operator(op.to_string()),
+};
+
+// Types
+
+TypeBinding: TypeBinding =
+    <id: Ident> <params: Ident*> "=" <body: Type> =>
+        TypeBinding::new(id, Type::Abs(params, Box::new(body)));
+
+FieldType: (Ident, Type) =
+    <id: Ident> ":" <ty: Type> =>
+        (id, ty);
+
+AtomicType: Type = {
+    Ident =>
+        Type::Ident(<>),
+
+    "{" <fields: Comma<FieldType>> "}" =>
+        Type::Record(fields),
+
+    "(" <elems: Comma<Type>> ")" =>
+        if elems.len() == 1 {
+            elems.into_iter().next().unwrap()
+        } else {
+            Type::tuple(elems)
+        },
+};
+
+AppType = {
+    AtomicType,
+
+    <ty: AtomicType> <args: AtomicType+> =>
+        Type::App(Box::new(ty), args),
+};
+
+InfixType = {
+    AppType,
+
+    <lhs: InfixType> <op: "Operator"> <rhs: AppType> =>
+        Type::Infix(Box::new(lhs), op.to_string(), Box::new(rhs)),
+};
+
+pub Type = {
+    InfixType,
+};
+
+// Patterns
+
+FieldPattern : (Ident, Pattern) = {
+    <id: Ident> "=" <pat: Pattern> => (id, pat),
+    <id: Ident> => (id.clone(), Pattern::Ident(id)),
+};
+
+AtomicPattern: Pattern = {
+    Ident =>
+        Pattern::Ident(<>),
+
+    "{" <fields: Comma<FieldPattern>> "}" =>
+        Pattern::Record(fields),
+
+    "(" <elems: Comma<Pattern>> ")" =>
+        if elems.len() == 1 {
+            elems.into_iter().next().unwrap()
+        } else {
+            Pattern::tuple(elems)
+        },
+};
+
+Pattern = {
+    AtomicPattern,
+
+    <id: Ident> <args: AtomicPattern+> =>
+        Pattern::Ctor(id, args),
+};
+
+// Expressions
+
+pub Literal: Literal = {
+    "StringLiteral" => Literal::String(<>),
+    "CharLiteral" => Literal::Char(<>),
+    "IntLiteral" => Literal::Int(<>),
+    "ByteLiteral" => Literal::Byte(<>),
+    "FloatLiteral" => Literal::Float(<>),
+};
+
+MatchArm: MatchArm =
+    "|" <pat: Pattern> "->" <expr: Expr> =>
+        MatchArm::new(pat, expr);
+
+FieldExpr: (Ident, Expr) =
+    <id: Ident> "=" <body: Expr> =>
+        (id, body);
+
+ValueBinding: ValueBinding = {
+    <id: Ident> <args: AtomicPattern+> "=" <body: Expr> =>
+        ValueBinding::new(Pattern::Ident(id), Expr::Lambda(args, Box::new(body))),
+
+    <pat: AtomicPattern> "=" <body: Expr> =>
+        ValueBinding::new(pat, body),
+};
+
+AtomicExpr: Expr = {
+    Ident =>
+        Expr::Ident(<>),
+
+    Literal =>
+        Expr::Literal(<>),
+
+    "(" "." <id: Ident> ")" =>
+        Expr::Getter(id),
```

Comment:
I think it's useful enough. Might want to add it in a follow up PR instead however